src/sock.c

```diff
@@ -282,7 +282,7 @@ static void read_conn(struct mg_connection *c) {
         if (!ioalloc(c, &c->rtls)) return;
         n = recv_raw(c, (char *) &c->rtls.buf[c->rtls.len],
                      c->rtls.size - c->rtls.len);
-        if (n == MG_IO_ERR && c->rtls.len == 0) {
```

Comment:
We've been going back and forth with this, breaking it several times.
Is there an actual reason to do this ? We've already been there and it didn't work, I think because there can be outstanding data in the buffer and TLS processed a whole record and hence not pending, so we never deliver the last chunk.

In reverse chronological order:
![Screenshot from 2024-07-26 16-40-59](https://github.com/user-attachments/assets/f7f4c693-e6f7-4064-abc3-801cff877aaa)
![Screenshot from 2024-07-26 16-41-45](https://github.com/user-attachments/assets/e55bbaab-dfbd-426b-87b5-426110addc15)
![Screenshot from 2024-07-26 16-42-35](https://github.com/user-attachments/assets/7d6062b5-e702-4417-8b86-4d54a6826ba1)
![Screenshot from 2024-07-26 16-43-20](https://github.com/user-attachments/assets/59a12243-a248-42df-b195-10fa8c774877)
![Screenshot from 2024-07-26 16-44-39](https://github.com/user-attachments/assets/9de28228-680a-4fc1-8807-2bb5f0ae3d0f)

Reply:
Yes there was a reason for that.
I started a HTTPS server on port 8443
And ran http client against it.
rtls buffered plain text HTTP request, thinking it is a TLS record, and thinking is was incomplete, so a connection was waiting for more.
I stopped HTTP client , but on the server side, a connection was still lingering in the infinite "read loop" .

Reply:
Yeah, I imagined something like that could happen... Then this should be the code, but for some reason we changed it, as we already changed the very one you just merged.
![image](https://github.com/user-attachments/assets/5d576df3-8241-44ab-bc75-7623a0367b62)
I'll revisit those issues on Monday and we can chat having data to support our brainstorming. WDYT ?

Reply:
Looks like that is a new corner case.
Your code will break some stuff.
We've been here
https://github.com/cesanta/mongoose/issues/2603#issuecomment-1938244497
https://github.com/cesanta/mongoose/issues/2719#issuecomment-2072530370
https://github.com/cesanta/mongoose/pull/2720
We should fix this without breaking those cases again.
I would revert that line back and probably re-think the whole scenario. As it is now, we have a regression and we'll probably have jcorporation posting an issue soon.

Reply:
@cpq

Reply:
IIUC, the problem is that non-TLS data in the buffer prevents proper closure... Is it right ?
Then, once we get IO_ERR there's nothing we can do after we deliver that data to TLS; so, what if we see IO_ERR, see data in rtls, remember that condition, keep going, deliver to TLS. Next time, we see we've been there and then close. WDYT ?
Something like that won't break those cases mentioned above.
After #2720, mg_tls_pending() or data in rtls causes the connection to be pending, that guarantees we'll arrive here, so we just need to remember we've already seen IO_ERR so if there is still outstanding data, whatever was there couldn't be processed by TLS. (Could this scenario also happen if a connection loss happens in the middle of a TLS block ? In such a case, asking for tls_pending won't help, but a fix like the one I'm proposing will work.)

Reply:
Sergio, your reasoning seems correct to - please go ahead and push a fix